### ISSUE TITLE
Return type of constructor defaults to class if typed with Union

### ISSUE THREAD

**Issue by matiusz** [microsoft/pyright#5986]
If class' `__new__` method is typed with a single class, Pyright correctly accepts changed type, eg.:

``` Python
from typing import Union
class Foo1:
    baz: int
class Custom:
    def __new__(cls, *args, **kwargs) -> Foo1:
        return super().__new__(cls)
x = Custom()
x.baz
```
If the typing is more complex though, eg. using an Union, the type of constructor defaults to the class the constructor belongs to.
``` Python
from __future__ import annotations
from typing import Union
class Foo1:
    baz: int
class Foo2:
    bar: str
class Custom:
    def __new__(cls, *args, **kwargs) -> Foo1 | Foo2:
        return super().__new__(cls)
    def __call__(self) -> Foo1 | Foo2:
        pass
x = Custom()
x.bar
x.baz
x().bar
x().baz
```

![image](https://github.com/microsoft/pyright/assets/37378034/352fec43-dd3b-4542-a91a-2c5fd84c12c9)

The expected behaviour is to have return value of constructor consistent with `__new__` even if typed with something like Union.

The issue started occuring between version 2023.5.30 and 2023.5.40 of Pylance, which includes changes from versions 1.1.308 and 1.1.309 of Pyright, most likely due to this commit: https://github.com/microsoft/pyright/commit/773b61942353b2444acdbb42ca098138f7f63d9f

**VS Code extension or command-line**
I'm using Pyright integrated with v2023.9.10 of Pylance.

**Comment by erictraut** [comment]
The type evaluation logic for constructor calls was largely rewritten in 1.1.309 to better reflect the behavior of the runtime. Constructor call type evaluation is complex because it involves the evaluation of multiple calls — the `__call__` method on the metaclass, the `__new__` method and the `__init__` method, and these individual calls can include conflicting type information. It's not possible for a static analyzer to determine the actual intent of the developer in all cases, so heuristics become necessary.

For an overview of how pyright handles constructor calls, refer to [this description](https://microsoft.github.io/pyright/#/mypy-comparison?id=constructor-calls). Note that other type checkers like mypy completely ignore the case where `__new__` returns a type other than the normal. Pyright attempts to handle the (still unusual) case where the return type is an instance of a different class or the same class but with different type arguments. In other words, pyright already goes significantly further than other type checkers in handling unusual cases.

What you're doing here by returning a union type from `__new__` falls far outside of the norm. This case causes one of pyright's heuristics to fail. It's not clear how this case should be handled in the presence of an `__init__` method (which your example admittedly doesn't include, but it's something that a type checker would need to deal with if it were to allow union types to be returned from `__new__`).

The constructor code is already very complex, full of heuristics, and prone to breaking existing code when these heuristics are modified. For these reasons, I'm not inclined to add support for this edge case. You'll need to find some workaround for your code.

We may be open to revisiting this decision if we receive signal from enough pyright users that this use case is more common than we suspect.
